File: ruby/grains/grains.rb
# frozen_string_literal: true

module Grains
  def self.square(square)
    unless (1..64).include?(square)
      raise ArgumentError, 'Square must be between 1 and 64'
    end

    2**(square - 1)
  end

  def self.total
    (0...64).sum { |x| 2**x }
  end
end

File: java/grains/Grains.java
import java.math.BigInteger;

class Grains {

    BigInteger grainsOnSquare(final int square) {
        if (!(1 <= square && square <= 64)) {
            throw new IllegalArgumentException("square must be between 1 and 64");
        }
        return BigInteger.ONE.shiftLeft(square - 1);

    }

    BigInteger grainsOnBoard() {
        BigInteger total = BigInteger.ZERO;
        for (int square = 1; square <= 64; square++) {
            total = total.add(grainsOnSquare(square));
        }
        return total;
    }

}
